canlink/busmanager.go

```diff
@@ -0,0 +1,242 @@
+//	 package main
+//
+//	 import (
+//	   "context"
+//	   "fmt"
+//	   "time"
+//
+//	   "go.einride.tech/can/pkg/socketcan"
+//	   "go.uber.org/zap"
+//	 )
+//
+//	 func main () {
+//	   ctx := context.Background()
+//
+//	   loggerConfig := zap.NewDevelopmentConfig()
+//	   logger, err := loggerConfig.Build()
+//
+//		 // Create a network connection for vcan0
+//		 conn, err := socketcan.DialContext(context.Background(), "can", "vcan0")
+//		 if err != nil {
+//		   return
+//		 }
+//
+//	   manager := canlink.NewBusManager(logger, conn)
+//	   handler = NewHandler(...)
+//
+//	   broadcast, transmit := manager.Register(handler)
+//
+//	   handler.Handle(...)
+//
+//	   manager.Start(ctx)
+//
+//	   ...
+//
+//	   manager.Stop()
+//	   manager.Close()
+type BusManager struct {
+	broadcastChan map[Handler]chan TimestampedFrame
+	incomingChan  chan TimestampedFrame
+
+	receiver    *socketcan.Receiver
+	transmitter *socketcan.Transmitter
+
+	l         *zap.Logger
+	stop      chan struct{}
+	isRunning bool
+	mu        sync.Mutex
+}
+
+// NewBusManager returns a BusManager object.
+//
+// The network connection is injected into the BusManager
+// and provides the interface for a single bus.
+//
+// See usage example.
+func NewBusManager(l *zap.Logger, conn *net.Conn) *BusManager {
+	busManager := &BusManager{
+		l: l.Named("Bus Manager"),
+
+		broadcastChan: make(map[Handler]chan TimestampedFrame),
+		incomingChan:  make(chan TimestampedFrame, _channelBufferLength),
+
+		receiver:    socketcan.NewReceiver(*conn),
+		transmitter: socketcan.NewTransmitter(*conn),
+	}
+
+	return busManager
+}
+
+// Register a Handler with the BusManager.
+//
+// Register returns two channels for the specified Handler.
+// The broadcast channel is a stream of traffic received from
+// the bus. The incoming channel enables the Handler to write
+// a frame to the bus.
+//
+// The channels operate on a TimestampedFrame object.
+func (b *BusManager) Register(
+	handler Handler,
+) (chan TimestampedFrame, chan TimestampedFrame) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	b.l.Info("registering handler")
+
+	subscription := make(chan TimestampedFrame, _channelBufferLength)
+	b.broadcastChan[handler] = subscription
+
+	b.l.Info("registered handler")
+
+	return subscription, b.incomingChan
+}
+
+// Unregister a Handler from the BusManager.
+//
+// Deletes the broadcast channel that was previously
+// provided from BusManager.
+func (b *BusManager) Unregister(handler *Handler) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	delete(b.broadcastChan, *handler)
+}
+
+// Start the traffic broadcast and incoming frame listener
+// for each of the registered handlers.
+//
+// The broadcast stream will begin. If handlers cannot keep up
+// with the broadcast, frames for that handler will be dropped.
+//
+// The handlers can now send CAN frames to the BusManager to be
+// transmitted out onto the bus.
+func (b *BusManager) Start(ctx context.Context) {
+	b.l.Info("start broadcast and incoming")
```

Comment:
replace "incoming" with "process incoming" or some other phrase that includes a verb. same with message in `.Stop`